spylon/spark/progress.py

```diff
@@ -0,0 +1,146 @@
+# INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+# WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from __future__ import print_function, absolute_import
+
+from collections import defaultdict
+import sys
+import datetime
+import time
+import threading
+
+
+def pretty_time_delta(td):
+    """A representation for timedelta
+
+    Parameters
+    ----------
+    td : :class:`datetime.timedelta`
+
+    Returns
+    -------
+    pretty_formatted_datetime : str
+
+    """
+    seconds = td.total_seconds()
+    sign_string = '-' if seconds < 0 else ''
+    seconds = abs(int(seconds))
+    days, seconds = divmod(seconds, 86400)
+    hours, seconds = divmod(seconds, 3600)
+    minutes, seconds = divmod(seconds, 60)
+    d = dict(sign=sign_string, days=days, hours=hours, minutes=minutes, seconds=seconds)
+    if days > 0:
+        return '{sign}{days}d{hours:02d}h{minutes:02d}m:{seconds:02d}s'.format(**d)
+    elif hours > 0:
+        return '{sign}{hours:02d}h{minutes:02d}m:{seconds:02d}s'.format(**d)
+    elif minutes > 0:
+        return '{sign}{minutes:02d}m:{seconds:02d}s'.format(**d)
+    else:
+        return '{sign}{seconds:02d}s'.format(**d)
+
+
+def spark_progress_thread_worker(status, timedelta_formatter=pretty_time_delta, bar_width=20, sleep_time=0.5):
+    """
+
+    Parameters
+    ----------
+    status : :class:`pyspark.status.StatusTracker`
+    timedelta_formatter : function
+        callable that converts a timedelta to a string.
+    bar_width : int
+        width of the progressbar to print out.
+    sleep_time : float
+        Frequency with which to poll spark for new progress information.
+
+    """
+    last_status = ''
+    start_times = defaultdict(datetime.datetime.now)
+    max_stage_id = 0
+
+    while True:
+        stage_ids = status.getActiveStageIds()
+        progressbar_list = []
+        # Only show first 3
+        for sid in stage_ids[:3]:
+            info = status.getStageInfo(sid)
+            if info:
+                td = datetime.datetime.now() - start_times[sid]
+                dur = timedelta_formatter(td)
+
+                percent = (info.numCompletedTasks * bar_width) // info.numTasks
+                bar = [' '] * bar_width
+                for i in range(bar_width):
+                    char = ' '
+                    if i < percent:
+                        char = '='
```

Comment:
Basically chose the symbols to use the exact same thing that the Scala console uses.